pkg/verifier/overlapped_blocks.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/improbable-eng/thanos/pkg/compact"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+)
+
+const OverlappedBlocksIssueID = "overlapped_blocks"
+
+// OverlappedBlocksIssue checks bucket for blocks with overlapped time ranges.
+// No repair is available for this issue.
+func OverlappedBlocksIssue() Issue {
+	return func(ctx context.Context, logger log.Logger, bkt objstore.Bucket, repair bool) error {
+		level.Info(logger).Log("msg", "started verifying issue", "with-repair", repair, "issue", OverlappedBlocksIssueID)
+
+		metas := map[string][]block.Meta{}
+		err := block.Foreach(ctx, bkt, func(id ulid.ULID) error {
+			m, err := block.DownloadMeta(ctx, bkt, id)
+			if err != nil {
+				return err
+			}
+
+			metas[compact.GroupKey(m)] = append(metas[compact.GroupKey(m)], m)
+			return nil
+		})
+		if err != nil {
+			return errors.Wrap(err, OverlappedBlocksIssueID)
+		}
+
+		var overlappedBlocks [][]block.Meta
+		for _, groupMetas := range metas {
+			overlap := findOverlappedBlocks(groupMetas)
+			if len(overlap) > 0 {
+				overlappedBlocks = append(overlappedBlocks, overlap...)
+			}
+		}
+
+		if len(overlappedBlocks) == 0 {
+			// All good.
+			return nil
+		}
+
+		for _, o := range overlappedBlocks {
+			var info []string
+			for _, m := range o {
+				r := time.Duration((m.MaxTime-m.MinTime)/1000) * time.Second
+				info = append(info, fmt.Sprintf(
+					"[id: %s mint: %d maxt: %d <%s>]",
+					m.ULID.String(),
+					m.MinTime,
+					m.MaxTime,
+					r.String(),
+				))
+			}
+			level.Warn(logger).Log("msg", "found overlapped blocks", "group", compact.GroupKey(o[0]), "blocks", strings.Join(info, ","))
+		}
+
+		if repair {
+			level.Warn(logger).Log("msg", "repair is not implemented for this issue", "issue", OverlappedBlocksIssueID)
+		}
+		return nil
+	}
+}
+
+type timestampToMeta struct {
+	meta *block.Meta
+	t    int64
+}
+
+func findOverlappedBlocks(metas []block.Meta) (overlappedBlocks [][]block.Meta) {
+	if len(metas) == 0 {
+		return overlappedBlocks
+	}
+
+	// Add both edge timestamps for each meta.
+	var timestamps []timestampToMeta
+	for _, m := range metas {
+		copy := m
+		timestamps = append(timestamps, timestampToMeta{meta: &copy, t: m.MinTime}, timestampToMeta{meta: &copy, t: m.MaxTime})
+	}
+	sort.Slice(timestamps, func(i int, j int) bool {
+		return timestamps[i].t < timestamps[j].t
+	})
+
+	// Blocks which we started but not yet finished.
+	var rollingBlocks = map[ulid.ULID]*block.Meta{}
+	for _, currT := range timestamps {
+		_, ok := rollingBlocks[currT.meta.ULID]
+		if ok {
+			// Let's finish current block if rolling.
+			delete(rollingBlocks, currT.meta.ULID)
+			continue
+		}
+
+		// New block, let's see if it's overlapping with anything rolling.
+		if len(rollingBlocks) > 0 {
+			var overlaps []block.Meta
+			for _, r := range rollingBlocks {
+				// It can happen that blocks started or finished in the same time. That's not a collision.
```

Comment:
When can that happen? Cannot think of a reason that would trigger it.

Reply:
you just mentioned the reason -> not stable sort.

But maybe just my comment is confusing. I meant here exactly that problem. It can happen that we have two blocks - one finished at A, second started at B. We should not treat that as overlap here.

Reply:
Second reason is that given meta can be not sorted anyway.